### PR TITLE
Add interface schema common and ethernet properties

### PR DESCRIPTION
**Pull request by EdDev**
This PR is dependent on PR#5 (https://github.com/nmstate/nmstate/pull/5)

### CODE REVIEW

File: libnmstate/schemas/interfaces.json [hunk 1]
```diff
@@ -8,23 +8,85 @@
...
+                    "minimum": 0
+                }
+            },
+            "required": ["name"]
+        },
+        "interface-ro": {
+            "title": "Interface",
+            "type": "object",
+            "properties": {
+                "if-index": { "type": "integer", "minimum": 0, "readOnly": true },
```
[the remark this answers is not in the document]
Comment (EdDev) [reply libnmstate/schemas/interfaces.json]: There are too many of them, so I prefer to handle the documentation (and probably other stuff) in future patches.

File: libnmstate/schemas/interfaces.json [hunk 2]
```diff
@@ -8,23 +8,85 @@
...
+                ]
             }
         }
     },
-    "required": ["interfaces"]
+    "required": ["interfaces"],
+
+    "definitions": {
+        "interface-rw": {
+            "title": "Interface",
```
[the remark this answers is not in the document]
Comment (EdDev) [reply libnmstate/schemas/interfaces.json]: json-schema defines all properties as RW unless marked as readOnly. I expect the doc generators that process a schema to base their output based on this definition.

File: libnmstate/schemas/interfaces.json [hunk 3]
```diff
@@ -8,23 +8,85 @@
...
+            "type": "object",
+            "properties": {
+                "name": { "type": "string" },
+                "description": { "type": "string" },
+                "type": {
+                    "type": "string",
+                    "enum": ["ethernet", "bond", "vlan"]
```
[the remark this answers is not in the document]
Comment (EdDev) [reply libnmstate/schemas/interfaces.json]: It is handled in the ethernet commit.

File: libnmstate/schemas/interfaces.json [hunk 4]
```diff
@@ -8,23 +8,85 @@
...
+            "required": ["name"]
+        },
+        "interface-ro": {
+            "title": "Interface",
+            "type": "object",
+            "properties": {
+                "if-index": { "type": "integer", "minimum": 0, "readOnly": true },
+                "admin-status": { "allOf": [ { "$ref": "#/definitions/values/status" }, { "readOnly": true } ] },
+                "link-status": { "allOf": [ { "$ref": "#/definitions/values/status" }, { "readOnly": true } ] },
+                "phys-address": { "allOf": [ { "$ref": "#/definitions/values/mac-address" }, { "readOnly": true } ] },
```
[the remark this answers is not in the document]
Comment (EdDev) [reply libnmstate/schemas/interfaces.json]: It is based on the agreement in the design document. I tried to follow existing naming from standards.

File: libnmstate/schemas/interfaces.json [hunk 5]
```diff
@@ -10,21 +10,89 @@
...
-                    "name": { "type": "string" },
-                    "description": { "type": "string" },
-                    "type": {
-                        "type": "string",
-                        "enum": ["ethernet", "bond", "vlan"]
-                    },
-                    "state": {
-                        "type": "string",
-                        "enum": ["absent", "up", "down", "unknown"]
+                "oneOf": [
```
Comment (phoracek) [inline libnmstate/schemas/interfaces.json]: why do use oneOf with only one item?

File: libnmstate/schemas/interfaces.json [hunk 6]
```diff
@@ -10,21 +10,89 @@
...
+                                    }
+                                }
+                            },
+                            { "$ref": "#/definitions/interface-rw" },
+                            { "$ref": "#/definitions/interface-ro" }
+                        ]
                     }
-                },
+                ],
                 "required": ["name"]
```
Comment (phoracek) [inline libnmstate/schemas/interfaces.json]: you require name byt it is not a property of this object

File: libnmstate/schemas/interfaces.json [hunk 7]
```diff
@@ -10,21 +10,89 @@
...
+                "admin-status": { "allOf": [ { "$ref": "#/definitions/values/status" }, { "readOnly": true } ] },
+                "link-status": { "allOf": [ { "$ref": "#/definitions/values/status" }, { "readOnly": true } ] },
+                "phys-address": { "allOf": [ { "$ref": "#/definitions/values/mac-address" }, { "readOnly": true } ] },
+                "higher-layer-if": { "type": "string", "readOnly": true },
+                "lower-layer-if": { "type": "string", "readOnly": true },
+                "statistics": {
+                    "title": "Interface statistics",
+                    "type": "object",
+                    "properties": {
+                        "in-octets": { "type": "integer", "minimum": 0 },
```
Comment (phoracek) [inline libnmstate/schemas/interfaces.json]: could you please sort them alphabetically?

File: libnmstate/schemas/interfaces.json [hunk 8]
```diff
@@ -10,21 +10,89 @@
...
+                        "in-errors": { "type": "integer", "minimum": 0 },
+                        "out-octets": { "type": "integer", "minimum": 0 },
+                        "out-unicast-pkts": { "type": "integer", "minimum": 0 },
+                        "out-broadcast-pkts": { "type": "integer", "minimum": 0 },
+                        "out-multicast-pkts": { "type": "integer", "minimum": 0 },
+                        "out-discards": { "type": "integer", "minimum": 0 },
+                        "out-errors": { "type": "integer", "minimum": 0 }
+                    },
+                    "readOnly": true,
+                    "required": ["in-octets", "in-unicast-pkts", "in-broadcast-pkts", "in-multicast-pkts", "in-discards", "in-errors", "out-octets", "out-unicast-pkts", "out-broadcast-pkts", "out-multicast-pkts", "out-discards", "out-errors"]
```
Comment (phoracek) [inline libnmstate/schemas/interfaces.json]: could you please write each requirement on one line? so we can easily sort them?

File: libnmstate/schemas/interfaces.json [hunk 9]
```diff
@@ -10,21 +10,89 @@
...
+                }
+            }
+        },
+        "values": {
+            "status": {
```
Comment (phoracek) [inline libnmstate/schemas/interfaces.json]: could you please rename it or add a description? it confuses me that we don't use the same type for rw state

File: libnmstate/schemas/interfaces.json (hunk 5 above)
Comment (phoracek) [reply libnmstate/schemas/interfaces.json]: I can see it in the following commit

File: tests/lib/schema_validation_test.py [hunk 10]
```diff
@@ -38,13 +38,17 @@
             'link-speed': 1000,
```
Comment (phoracek) [inline tests/lib/schema_validation_test.py]: please add one test for ethernet and one for common (bond or vlan)

File: libnmstate/schemas/interfaces.json [hunk 11]
```diff
@@ -94,5 +101,41 @@
...
+                    "oneOf": [
+                        { "$ref": "#/__definitions__/auto-false" },
+                        { "$ref": "#/__definitions__/auto-true" }
+                    ]
+                }
+            ],
+            "required": [ "type", "auto-negotiation" ]
+        }
+    },
+    "__definitions__": {
```
Comment (phoracek) [inline libnmstate/schemas/interfaces.json]: please add an inline comment what we keep in __definitions__

File: libnmstate/schemas/interfaces.json [hunk 12]
```diff
@@ -94,5 +101,41 @@
...
+                    }
+                },
+                {
+                    "oneOf": [
+                        { "$ref": "#/__definitions__/auto-false" },
+                        { "$ref": "#/__definitions__/auto-true" }
+                    ]
+                }
+            ],
+            "required": [ "type", "auto-negotiation" ]
```
Comment (phoracek) [inline libnmstate/schemas/interfaces.json]: i don't like the fact that we need to require something that is hidden in reference. is there a way around? if not, please keep one requirement per line and add inline comment to where is it required from.

File: libnmstate/schemas/interfaces.json [hunk 13]
```diff
@@ -94,5 +101,41 @@
...
+        }
+    },
+    "__definitions__": {
+        "auto-false": {
```
Comment (phoracek) [inline libnmstate/schemas/interfaces.json]: what is this? could you comment on it?

File: libnmstate/schemas/interfaces.json (hunk 6 above)
Comment (EdDev) [reply libnmstate/schemas/interfaces.json]: It needs to be located in the same level as the "properties" and works on it.

File: libnmstate/schemas/interfaces.json (hunk 9 above)
Comment (EdDev) [reply libnmstate/schemas/interfaces.json]: These are general "variables" which usually appear in several places, so we define them once and reference to it. How is it related to RW exactly?

File: tests/lib/schema_validation_test.py (hunk 10 above)
Comment (EdDev) [reply tests/lib/schema_validation_test.py]: Ok, I will do it in the next commit.

File: libnmstate/schemas/interfaces.json (hunk 11 above)
Comment (EdDev) [reply libnmstate/schemas/interfaces.json]: As mentioned, comments are not supported.

File: libnmstate/schemas/interfaces.json (hunk 13 above)
Comment (phoracek) [reply libnmstate/schemas/interfaces.json]: So title or description?

Comment (EdDev) [reply libnmstate/schemas/interfaces.json]: That is used for the api documentation, not for the schema developer/editor clarity.

File: libnmstate/schemas/interfaces.json (hunk 9 above)
Comment (phoracek) [reply libnmstate/schemas/interfaces.json]: I want you to give "status" more explicit name so it is obvious why this "status" is not used for rw "state"

**Pull request merged**